flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java

```diff
@@ -0,0 +1,255 @@
+            LOG.warn(
+                "Unable to add splits back for subtask {} since it is not assigned any splits. Splits: {}",
+                subtaskId,
+                splits);
+            return;
+        }
+
+        for (KinesisShardSplit split : splits) {
+            splitAssignment.get(subtaskId).remove(split);
+            assignedSplits.remove(split);
+            unassignedSplits.add(split);
+        }
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        splitAssignment.putIfAbsent(subtaskId, new HashSet<>());
+    }
+
+    @Override
+    public KinesisStreamsSourceEnumeratorState snapshotState(long checkpointId) throws Exception {
+        return new KinesisStreamsSourceEnumeratorState(completedShards, unassignedSplits, lastSeenShardId);
+    }
+
+    @Override
+    public void close() throws IOException {
+    }
+
+    private List<KinesisShardSplit> initialDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        return mapToSplits(shards, false);
+    }
+
+    private List<KinesisShardSplit> periodicallyDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        // Any shard discovered after the initial startup should be read from the start, since they
+        // come from resharding
+        return mapToSplits(shards, true);
+    }
+
+    private List<KinesisShardSplit> mapToSplits(List<Shard> shards, boolean shouldReadFromStart) {
+        InitialPosition startingPosition =
+            shouldReadFromStart
+                ? InitialPosition.TRIM_HORIZON
+                : InitialPosition.valueOf(
+                consumerConfig
+                    .getOrDefault(
+                        STREAM_INITIAL_POSITION,
+                        DEFAULT_STREAM_INITIAL_POSITION)
+                    .toString());
+        long startingTimestamp = 0;
+        switch (startingPosition) {
+            case LATEST:
+                startingTimestamp = Instant.now().toEpochMilli();
+                break;
+            case AT_TIMESTAMP:
+                startingTimestamp = parseStreamTimestampStartingPosition(consumerConfig).getTime();
+                break;
+            case TRIM_HORIZON:
+            default:
+                // Since we are reading from TRIM_HORIZON, starting time epoch millis can be 0
+        }
+
+        List<KinesisShardSplit> splits = new ArrayList<>();
+        for (Shard shard : shards) {
+            splits.add(
+                new KinesisShardSplit(
+                    streamArn,
+                    shard.shardId(),
+                    startingPosition.toString(),
+                    startingTimestamp));
+        }
+
+        return splits;
+    }
+
+    private void assignSplits(List<KinesisShardSplit> discoveredSplits, Throwable t) {
+        if (t != null) {
+            throw new FlinkRuntimeException("Failed to list shards.", t);
```

Comment:
Can more specific exception be used here?

Reply:
Great shout! Added a `KinesisStreamsSourceException`